lib/tests/formulae.rb

```diff
@@ -251,8 +257,9 @@ def formula!(formula_name, args:)
         deps = []
         reqs = []
 
+        build_bottle = !formula.bottle_disabled? && !args.build_from_source? && all_bottled_deps?(formula)
         fetch_args = [formula_name]
-        fetch_args << "--build-bottle" if !formula.bottle_disabled? && !args.build_from_source?
+        fetch_args << "--build-bottle" if build_bottle
```

Comment:
Yes. Similarly with doing dependent tests in a separate workflow with the goal of parallelising them (possibly even on GitHub runners).

Reply:
Just in case I misunderstood: is this a future ambition or something you're planning to implement? It will require a fair amount of changes on the workflow side.

Reply:
Future ambition.

Reply:
Oh ok then. To take a step back: this PR is just about attempting to build from source always but only actually build a bottle if the dependencies are bottled?

If so, LGTM. Maybe it should use an explicit `--build-from-source` though.

Reply:
> Oh ok then. To take a step back: this PR is just about attempting to build from source always but only actually build a bottle if the dependencies are bottled?

Yes.

> If so, LGTM. Maybe it should use an explicit `--build-from-source` though.

Hmmm, yea, probably. Will this just error out if we don't have `--build-from-source`?

Reply:
We unset `HOMEBREW_DEVELOPER` so yeah it should error I think if there's no bottle available.

Or it might at least get confused since there'll be a bottle block for the previous version.

Reply:
I think I'm fine with an error for now. At this point my most important concern is a `dispatch-build-bottle` being issued for a formula with an unbottled dependency.

Reply:
So no plans to enable Monterey for PRs then?

Reply:
I don't know. We're at ~55% bottle coverage. We have some important formulae with no bottles yet, though (`openjdk`, `apr`). I'm happy to enable it now. Or I'm also happy to bottle more formulae for another day or so to get it to somewhere closer to 70%.

Reply:
I mean as long as it's not 100%, we'll need to make sure this doesn't error.

Reply:
Ah yes, good point. I guess we want `--build-from-source` then.

Reply:
> We unset `HOMEBREW_DEVELOPER` so yeah it should error I think if there's no bottle available.

It behaves differently if `CI` is set.

Reply:
> So no plans to enable Monterey for PRs then?

I think we should enable ASAP but have it skip (without error) if not already bottled (like is done on ARM and Linux).

Reply:
Cool. I'll work on enabling this today or tomorrow; just trying to catch the last few stragglers from `brew unbottled` that I haven't attempted bottle jobs for yet..

Reply:
@carlocab sounds good!

Reply:
> > We unset `HOMEBREW_DEVELOPER` so yeah it should error I think if there's no bottle available.
> 
> It behaves differently if `CI` is set.

I don't see where that logic is but in any case the presence of the previous bottle block in a version bump PR is probably enough to cause issues.

Reply:
> I don't see where that logic is

I lied, I removed this recently as part of the Linuxbrew migration (as it used to apply always on macOS or when CI was enabled on Linux). Ignore me!